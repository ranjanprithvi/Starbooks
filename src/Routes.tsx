import React from "react";
import { Navigate, Route, Routes as ReactRoutes } from "react-router-dom";
import Books from "./components/Books/Books";
import BookDetails from "./components/Books/BookDetails";
import BookForm from "./components/Books/BookForm";
import NotFound from "./components/common/NotFound";
import Rentals from "./components/Rentals/Rentals";
import Users from "./components/Users/Users";
import RentalForm from "./components/Rentals/RentalForm";
import UserDetails from "./components/Users/UserDetails";
import UserForm from "./components/Users/UserForm";
import LoginForm from "./components/Users/LoginForm";
import ProtectedAdminComponent from "./components/common/ProtectedAdminComponent";
import ProtectedComponent from "./components/common/ProtectedComponent";
import AuthorGenreList from "./components/Authors/AuthorGenreList";
import Home from "./components/Home";

const Routes = ({ dataView }: { dataView?: string | undefined }) => {
    return (
        <ReactRoutes>
            <Route path="/" element={<Home />} />
            <Route path="/login" element={<LoginForm />} />
            <Route path="/books" element={<Books />} />
            <Route path="bookDetails/:id" element={<BookDetails />} />
            <Route
                path="books/:id"
                element={
                    <ProtectedAdminComponent>
                        <BookForm />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="users"
                element={
                    <ProtectedAdminComponent>
                        <Users dataView={dataView} />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="userDetails/:id"
                element={
                    <ProtectedComponent>
                        <UserDetails />
                    </ProtectedComponent>
                }
            />
            <Route
                path="users/:id"
                element={
                    <ProtectedAdminComponent>
                        <UserForm />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="rentals"
                element={
                    <ProtectedAdminComponent>
                        <Rentals />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="rentals/:id"
                element={
                    <ProtectedAdminComponent>
                        <RentalForm />
                    </ProtectedAdminComponent>
                }
            />
            <Route path="/authorGenreList" element={<AuthorGenreList />} />
            <Route path="/not-found" element={<NotFound />} />
            <Route path="*" element={<Navigate to="/not-found" replace />} />
        </ReactRoutes>
    );
};

export default Routes;
