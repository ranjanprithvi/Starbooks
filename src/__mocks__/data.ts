export const bookObj = {
    _id: "123",
    title: "test",
    author: { _id: "1", name: "author1" },
    genre: { _id: "1", name: "genre1" },
    yearPublished: 2021,
    rating: 5,
    coverImage:
        "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1681496408i/58536026.jpg",
    description: "test",
    numberInStock: 1,
};
