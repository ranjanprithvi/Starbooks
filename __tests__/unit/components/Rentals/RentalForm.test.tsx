import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import apiClient from "../../../../src/services/api-client";
import React from "react";
import { bookObj, booksArray, usersArray } from "../../../../src/data/mockData";
import { MemoryRouter } from "react-router-dom";
import Routes from "../../../../src/Routes";
import _ from "lodash";

jest.mock("../../../../src/hooks/useBooks");
jest.mock("../../../../src/hooks/useUsers");

jest.mock("../../../../src/components/common/ProtectedAdminComponent");

//mock apiClient
jest.mock("../../../../src/services/api-client");
const mApiClient = apiClient as jest.Mocked<typeof apiClient>;
mApiClient.post.mockResolvedValue({ data: bookObj });

//mock useToast
const mToast = jest.fn();
jest.mock("@chakra-ui/react", () => ({
    ...(jest.requireActual("@chakra-ui/react") as any),
    useToast: () => mToast,
}));

//mock useNavigate
const mNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
    ...(jest.requireActual("react-router-dom") as any),
    useNavigate: () => mNavigate,
}));

describe("RentalForm", () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it("renders", async () => {
        const tree = renderer
            .create(
                <MemoryRouter initialEntries={[`/rentals/new`]}>
                    <Routes />
                </MemoryRouter>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("renders with none of the fields populated when creating a rental", async () => {
        render(
            <MemoryRouter initialEntries={[`/rentals/new`]}>
                <Routes />
            </MemoryRouter>
        );
        expect(screen.getByLabelText(/book/i)).toHaveValue("");
        expect(screen.getByLabelText(/user/i)).toHaveValue("");
    });

    it("does not submit when book is not selected", async () => {
        render(
            <MemoryRouter initialEntries={[`/rentals/new`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            fireEvent.change(screen.getByRole("combobox", { name: /user/i }), {
                target: { value: usersArray[0]._id },
            });
            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.post).not.toHaveBeenCalled();
        expect(screen.getByText(/book is required/i)).toBeInTheDocument();
    });

    it("does not submit when user is not selected", async () => {
        render(
            <MemoryRouter initialEntries={[`/rentals/new`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            fireEvent.change(screen.getByRole("combobox", { name: /book/i }), {
                target: { value: booksArray[0]._id },
            });
            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.post).not.toHaveBeenCalled();
        expect(screen.getByText(/user is required/i)).toBeInTheDocument();
    });

    it("submits with the correct data when both book and user are selected", async () => {
        render(
            <MemoryRouter initialEntries={[`/rentals/new`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            fireEvent.change(screen.getByRole("combobox", { name: /book/i }), {
                target: { value: booksArray[0]._id },
            });
            fireEvent.change(screen.getByRole("combobox", { name: /user/i }), {
                target: { value: usersArray[0]._id },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.post).toHaveBeenCalledWith("/rentals", {
            user: usersArray[0]._id,
            book: booksArray[0]._id,
        });
    });

    it("displays error message in case of network error", async () => {
        const message = "Sorry. Something went wrong";
        mApiClient.post.mockRejectedValueOnce({});
        // process.env.VITE_BACKEND_URL = "http://invalidUrl";
        render(
            <MemoryRouter initialEntries={[`/rentals/new`]}>
                <Routes />
            </MemoryRouter>
        );

        await act(async () => {
            fireEvent.change(screen.getByRole("combobox", { name: /book/i }), {
                target: { value: booksArray[0]._id },
            });
            fireEvent.change(screen.getByRole("combobox", { name: /user/i }), {
                target: { value: usersArray[0]._id },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });
        expect(mToast).toHaveBeenCalled();
    });

    it("navigates to previous screen when canceled", async () => {
        render(
            <MemoryRouter initialEntries={[`/rentals/new`]}>
                <Routes />
            </MemoryRouter>
        );

        const cancelButton = screen.getByText(/cancel/i);
        cancelButton.click();
        expect(mNavigate).toHaveBeenCalledWith(-1);
    });
});
