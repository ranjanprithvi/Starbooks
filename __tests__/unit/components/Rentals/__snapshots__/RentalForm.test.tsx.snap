// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RentalForm renders 1`] = `
<div
  className="css-3xydd0"
>
  <div
    className="css-zmtksu"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="chakra-stack css-tl3ftk"
      >
        <h2
          className="chakra-heading css-r7q7qr"
        >
          New Rental
        </h2>
        <div
          className="chakra-form-control css-h82jag"
          role="group"
        >
          <div
            className="chakra-form-control css-0"
            role="group"
          >
            <label
              className="chakra-form__label css-uvho5g"
              htmlFor="book"
              id="field-:r1:-label"
            >
              Book
            </label>
            <div
              className="chakra-select__wrapper css-42b2qy"
            >
              <select
                className="chakra-select css-1ik61og"
                disabled={false}
                id="book"
                name="book"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                readOnly={false}
                required={false}
              >
                <option
                  value=""
                >
                  --Select book--
                </option>
                <option
                  disabled={true}
                  value="65987aa9e6b72d92ddcc1754"
                >
                  asdf
                </option>
                <option
                  disabled={false}
                  value="651becad552934bbabb7bdd5"
                >
                  The Body in the Library
                </option>
                <option
                  disabled={true}
                  value="651765abb66bbab77eb9ed5b"
                >
                  Murder on the Orient Express
                </option>
                <option
                  disabled={false}
                  value="651760180012d3fa25b3ddba"
                >
                  The Code of the Woosters
                </option>
                <option
                  disabled={false}
                  value="651486abc59681a0aa4771ed"
                >
                  The Kite Runner
                </option>
                <option
                  disabled={false}
                  value="651485f0c59681a0aa4771df"
                >
                  Inheritance
                </option>
                <option
                  disabled={false}
                  value="651452f1c59681a0aa477199"
                >
                  Brisinger
                </option>
                <option
                  disabled={false}
                  value="65145251c59681a0aa47718b"
                >
                  Eldest
                </option>
                <option
                  disabled={false}
                  value="65145107c59681a0aa477174"
                >
                  Charlie and the great glass elevator
                </option>
              </select>
              <div
                className="chakra-select__icon-wrapper css-162mkon"
              >
                <svg
                  aria-hidden={true}
                  className="chakra-select__icon"
                  focusable={false}
                  role="presentation"
                  style={
                    {
                      "color": "currentColor",
                      "height": "1em",
                      "width": "1em",
                    }
                  }
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          className="chakra-form-control css-h82jag"
          role="group"
        >
          <div
            className="chakra-form-control css-0"
            role="group"
          >
            <label
              className="chakra-form__label css-uvho5g"
              htmlFor="user"
              id="field-:r3:-label"
            >
              User
            </label>
            <div
              className="chakra-select__wrapper css-42b2qy"
            >
              <select
                className="chakra-select css-1ik61og"
                disabled={false}
                id="user"
                name="user"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                readOnly={false}
                required={false}
              >
                <option
                  value=""
                >
                  --Select user--
                </option>
                <option
                  disabled={false}
                  value="1"
                >
                  John
                </option>
                <option
                  disabled={false}
                  value="2"
                >
                  Jane
                </option>
                <option
                  disabled={false}
                  value="3"
                >
                  Jack
                </option>
              </select>
              <div
                className="chakra-select__icon-wrapper css-162mkon"
              >
                <svg
                  aria-hidden={true}
                  className="chakra-select__icon"
                  focusable={false}
                  role="presentation"
                  style={
                    {
                      "color": "currentColor",
                      "height": "1em",
                      "width": "1em",
                    }
                  }
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          className="chakra-stack css-mogqpi"
        >
          <button
            className="chakra-button css-4xx2wk"
            id="submit"
            type="submit"
          >
            Submit
          </button>
          <button
            className="chakra-button css-4xx2wk"
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
